src/Salsa.jl

```diff
@@ -46,13 +46,31 @@ const CallArgs = Tuple  # (Unused, but defined here for clarity. Could be used i
 # identifies requests for the cached value (ie accessing inputs or calling derived
 # functions). Instances of these key types are used to identify _which maps_ the remaining
 # arguments (as a CallArgs tuple) are the key to.
-# e.g. foo(component, 2,3,5) -> DerivedKey{Foo, (Int,Int,Int)}()
+# e.g. foo(component, 2,3,5) -> DerivedKey{Foo, (MyComponent,Int,Int,Int)}()
 abstract type AbstractKey end
-# To specify dependencies between Salsa computations, we use a tuple whose first element is a
-# key instance that specifies a call (above) and remaining elements are the arguments to
-# that call.
-# e.g. foo(component, 2,3,5) -> (DerivedKey{Foo, (Int,Int,Int)}(), 2,3,5)
-const DependencyKey = Tuple{AbstractKey, Vararg{Any}}
+# To specify dependencies between Salsa computations, we use a named tuple that stores
+# everything needed to rerun the computation: (key, function-args)
+# The `key` is the AbstractKey instance that specifies _which computation_ was performed
+# (above) and the `args` is a Tuple of the user-provided arguments to that call. If the
+# computation was a derived function, the first argument of the `args` tuple will be an
+# `AbstractComponent`, as is required for Derived functions.
+# Examples:
+#   foo(component,1,2,3) -> DependencyKey(key=DerivedKey{Foo, (MyComponent,Int,Int,Int)}(),
+#                                         args=(db, 1, 2, 3))
+#   db.map[1,2]    -> DependencyKey(key=InputKey{:map}(), args=(1, 2))
+Base.@kwdef struct DependencyKey{KT<:AbstractKey}
+    key::KT
+    args::Tuple
+end
```

Comment:
```suggestion
Base.@kwdef struct DependencyKey{KT<:AbstractKey, ARG_T<:Tuple}
    key::KT
    args::ARG_T
end
```

Isn't it better to have type argument for the `args` field? Otherwise, there's no chance that Julia can specialize on the type of arguments.

Reply:
Yeah, i think that's probably a good suggestion, thanks! :)

I'm not sure how much of a difference it'll make, because i think most of the times we use `args`, it's in generic code that won't be any different depending on the types. In fact we might even want to look into marking some of it as `@nospecialize`... But that's all future nitty-gritty optimization work. I'll make this change for now. Thanks! :)